src/dune/preprocess.ml

```diff
@@ -0,0 +1,251 @@
+    if pps = [] then No_preprocessing
+    else Pps {t with pps}
+  | No_preprocessing | Action _ | Future_syntax _ as t -> t
+
+let decode =
+  sum
+    [ ("no_preprocessing", return No_preprocessing)
+    ; ( "action"
+      , located Action_dune_lang.decode >>| fun (loc, x) -> Action (loc, x) )
+    ; ( "pps"
+      , let+ loc = loc
+        and+ pps, flags = Pps_and_flags.decode in
+        Pps { loc; pps; flags; staged = false } )
+    ; ( "staged_pps"
+      , let+ () = Dune_lang.Syntax.since Stanza.syntax (1, 1)
+        and+ loc = loc
+        and+ pps, flags = Pps_and_flags.decode in
+        Pps { loc; pps; flags; staged = true } )
+    ; ( "future_syntax"
+      , let+ () = Dune_lang.Syntax.since Stanza.syntax (1, 8)
+        and+ loc = loc in
+        Future_syntax loc )
+    ]
+
+let loc = function
+  | No_preprocessing -> None
+  | Action (loc, _)
+  | Pps { loc; _ }
+  | Future_syntax loc ->
+    Some loc
+
+let pps = function
+  | Pps { pps; _ } -> pps
+  | _ -> []
+
+module Without_future_syntax = struct
+  type 'a t =
+    | No_preprocessing
+    | Action of Loc.t * Action_dune_lang.t
+    | Pps of 'a Pps.t
+end
+
+module Pp_flag_consumer = struct
+  (* Compiler allows the output of [-pp] to be a binary AST. Merlin requires
+     that to be a text file instead. *)
+  type t =
+    | Compiler
+    | Merlin
+end
+
+let remove_future_syntax (t : 'a t) ~(for_ : Pp_flag_consumer.t) v :
+  'a Without_future_syntax.t =
+  match t with
+  | No_preprocessing -> No_preprocessing
+  | Action (loc, action) -> Action (loc, action)
+  | Pps pps -> Pps pps
+  | Future_syntax loc ->
+    if Ocaml_version.supports_let_syntax v then
+      No_preprocessing
+    else
+      Action
+        ( loc
+        , Run
+            ( String_with_vars.make_var loc "bin" ~payload:"ocaml-syntax-shims"
+            , ( match for_ with
+                | Compiler -> [ String_with_vars.make_text loc "-dump-ast" ]
+                | Merlin ->
+                  (* We generate a text file instead of AST. That gives you less
+                     precise locations, but at least Merlin doesn't fail
+                     outright.
+
+                     In general this hack should be applied to all -pp commands
+                     that might produce an AST, not just to Future_syntax. But
+                     doing so means we need to change dune language so the user
+                     can provide two versions of the command.
+
+                     Hopefully this will be fixed in merlin before that becomes
+                     a necessity. *)
+                  [] )
+              @ [ String_with_vars.make_var loc "input-file" ] ) )
+
+module Without_instrumentation = struct
+  type t = Loc.t * Lib_name.t
+
+  let compare_no_locs (_, x) (_, y) =
+    Lib_name.compare x y
+end
+
+module With_instrumentation = struct
+  type t =
+    | Ordinary of Without_instrumentation.t
+    | Instrumentation of (Loc.t * Lib_name.t)
```

Comment:
I would call this `Instrumentation_backend` to be fully explicit

Reply:
Yes, fixed.